File: app/src/main/java/de/bahnhoefe/deutschlands/bahnhofsfotos/model/Country.java
package de.bahnhoefe.deutschlands.bahnhofsfotos.model;

import androidx.annotation.NonNull;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Optional;

import lombok.Builder;
import lombok.EqualsAndHashCode;
import lombok.Value;

@Value
@Builder
@EqualsAndHashCode(onlyExplicitlyIncluded = true)
public class Country implements Serializable, Comparable<Country> {

    @EqualsAndHashCode.Include
    String code;

    String name;

    String email;

    String twitterTags;

    String timetableUrlTemplate;

    String overrideLicense;

    @Builder.Default
    List<ProviderApp> providerApps = new ArrayList<>();

    public static Optional<Country> getCountryByCode(Collection<Country> countries, String countryCode) {
        return countries.stream()
                .filter(country -> country.getCode().equals(countryCode))
                .findAny(); // get first country as fallback
    }

    public boolean hasTimetableUrlTemplate() {
        return timetableUrlTemplate != null && !timetableUrlTemplate.isEmpty();
    }

    public List<ProviderApp> getCompatibleProviderApps() {
        return ProviderApp.getCompatibleProviderApps(providerApps);
    }

    @Override
    @NonNull
    public String toString() {
        return name;
    }

    @Override
    public int compareTo(final Country o) {
        return name.compareTo(o.name);
    }

}

File: app/src/main/kotlin/de/bahnhoefe/deutschlands/bahnhofsfotos/model/Country.kt
package de.bahnhoefe.deutschlands.bahnhofsfotos.model

import java.io.Serializable
import java.util.Objects

data class Country @JvmOverloads constructor(
    val code: String,
    val name: String,
    val email: String? = null,
    val timetableUrlTemplate: String? = null,
    val overrideLicense: String? = null,
    val providerApps: List<ProviderApp> = listOf()
) : Serializable, Comparable<Country> {

    fun hasTimetableUrlTemplate(): Boolean {
        return !timetableUrlTemplate.isNullOrEmpty()
    }

    val compatibleProviderApps: List<ProviderApp>
        get() = getCompatibleProviderApps(providerApps)

    override fun toString(): String {
        return name
    }

    override fun compareTo(other: Country): Int {
        return name.compareTo(other.name)
    }

    override fun equals(other: Any?): Boolean {
        if (other === this) return true
        if (other !is Country) return false
        return Objects.equals(code, other.code)
    }

    override fun hashCode(): Int {
        return code.hashCode()
    }
}

fun getCountryByCode(countries: Collection<Country>, countryCode: String?) = countries
    .firstOrNull { country: Country -> country.code == countryCode }
